File: apps/desktop/src/lib/contracts/job.ts
import { z } from "zod";

export const jobModeSchema = z.enum(["copy", "move"]);
export const conflictStrategySchema = z.enum(["skip", "overwrite", "rename"]);
export const symlinkModeSchema = z.enum(["follow", "copyLink", "skip"]);

export const EXCLUDE_PRESETS = [
  { id: "node_modules", label: "node_modules" },
  { id: "git", label: ".git" },
  { id: "temp", label: "temp/tmp" },
  { id: "system", label: "Sistema" },
  { id: "build", label: "build/dist" },
] as const;

const isoDate = z.string().datetime().nullable().optional();

export const fileFilterSchema = z.object({
  includeExtensions: z.array(z.string()).default([]),
  excludePatterns: z.array(z.string()).default([]),
  excludePresetIds: z.array(z.string()).default([]),
  nameRegex: z.string().nullable().optional(),
  pathRegex: z.string().nullable().optional(),
  minSizeBytes: z.number().int().nonnegative().nullable().optional(),
  maxSizeBytes: z.number().int().nonnegative().nullable().optional(),
  maxDepth: z.number().int().nonnegative().nullable().optional(),
  modifiedAfter: isoDate,
  modifiedBefore: isoDate,
  createdAfter: isoDate,
  createdBefore: isoDate,
  olderThanDays: z.number().int().nonnegative().nullable().optional(),
  contentContains: z.string().nullable().optional(),
  contentMaxBytes: z.number().int().positive().default(5 * 1024 * 1024),
  contentExtensions: z.array(z.string()).default([
    ".txt", ".md", ".csv", ".json", ".log", ".xml", ".yaml", ".yml",
  ]),
  recursive: z.boolean().default(true),
  includeHidden: z.boolean().default(false),
  symlinkMode: symlinkModeSchema.default("follow"),
  skipEmptyFiles: z.boolean().default(false),
});

export const transferOptionsSchema = z.object({
  smartSync: z.boolean().default(false),
  strictHashSync: z.boolean().default(false),
  verifyAfterCopy: z.boolean().default(true),
  stopOnIntegrityError: z.boolean().default(false),
  encryptOutput: z.boolean().default(false),
  encryptPassword: z.string().nullable().optional(),
  rememberEncryptPassword: z.boolean().default(false),
  removeFilesAfterPack: z.boolean().default(false),
  packFilename: z.string().nullable().optional(),
});

export const scheduleRuleSchema = z.discriminatedUnion("kind", [
  z.object({ kind: z.literal("interval"), minutes: z.number().int().min(1) }),
  z.object({ kind: z.literal("daily"), hour: z.number().int().min(0).max(23), minute: z.number().int().min(0).max(59) }),
  z.object({
    kind: z.literal("weekly"),
    days: z.array(z.number().int().min(0).max(6)),
    hour: z.number().int().min(0).max(23),
    minute: z.number().int().min(0).max(59),
  }),
]);

export const scheduleConfigSchema = z.object({
  enabled: z.boolean(),
  timezone: z.string().default("local"),
  rule: scheduleRuleSchema,
});

export const scriptsConfigSchema = z.object({
  preScript: z.string().nullable().optional(),
  postScript: z.string().nullable().optional(),
  timeoutSecs: z.number().int().min(5).max(600).default(60),
});

export const notifyEventSchema = z.enum(["started", "completed", "failed"]);

export const notifyChannelSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("generic"),
    url: z.string().url(),
    headers: z.record(z.string()).default({}),
  }),
  z.object({ type: z.literal("discord"), webhookUrl: z.string().url() }),
  z.object({
    type: z.literal("telegram"),
    botToken: z.string().min(1),
    chatId: z.string().min(1),
    rememberToken: z.boolean().default(false),
  }),
]);

export const notifyConfigSchema = z.object({
  enabled: z.boolean().default(false),
  events: z.array(notifyEventSchema).default(["completed", "failed"]),
  channels: z.array(notifyChannelSchema).default([]),
});

export const jobSchema = z.object({
  id: z.string().uuid(),
  name: z.string().min(1).max(120),
  sourcePath: z.string(),
  targetPath: z.string(),
  mode: jobModeSchema,
  conflict: conflictStrategySchema,
  filters: fileFilterSchema.default({}),
  options: transferOptionsSchema.default({}),
  schedule: scheduleConfigSchema.nullable().optional(),
  scripts: scriptsConfigSchema.default({}),
  notify: notifyConfigSchema.default({}),
  enabled: z.boolean(),
  lastRun: z.string().datetime().nullable(),
  nextRun: z.string().datetime().nullable().optional(),
});

export const jobSummarySchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  sourcePath: z.string(),
  targetPath: z.string(),
  mode: jobModeSchema,
  enabled: z.boolean(),
  lastRun: z.string().datetime().nullable(),
  nextRun: z.string().datetime().nullable().optional(),
  scheduleEnabled: z.boolean().optional(),
});

export const activeExecutionSchema = z.object({
  executionId: z.string().uuid(),
  jobId: z.string().uuid(),
  jobName: z.string(),
  currentFile: z.string(),
  percent: z.number(),
  bytesPerSec: z.number(),
  recentLog: z.array(z.string()),
});

export const executionProgressSchema = activeExecutionSchema;

export const executionCompletedSchema = z.object({
  executionId: z.string().uuid(),
  jobId: z.string().uuid(),
  success: z.boolean(),
  processed: z.number().int(),
  failed: z.number().int(),
  errorMessage: z.string().nullable(),
});

export const simulationSampleSchema = z.object({
  source: z.string(),
  target: z.string(),
  action: z.string(),
});

export const simulationReportSchema = z.object({
  filesMatched: z.number().int(),
  filesSkipped: z.number().int(),
  sample: z.array(simulationSampleSchema),
  warnings: z.array(z.string()),
});

export type Job = z.infer<typeof jobSchema>;
export type JobSummary = z.infer<typeof jobSummarySchema>;
export type FileFilter = z.infer<typeof fileFilterSchema>;
export type TransferOptions = z.infer<typeof transferOptionsSchema>;
export type ScheduleConfig = z.infer<typeof scheduleConfigSchema>;
export type ScheduleRule = z.infer<typeof scheduleRuleSchema>;
export type ScriptsConfig = z.infer<typeof scriptsConfigSchema>;
export type NotifyConfig = z.infer<typeof notifyConfigSchema>;
export type NotifyChannel = z.infer<typeof notifyChannelSchema>;
export type SimulationReport = z.infer<typeof simulationReportSchema>;
export type ActiveExecution = z.infer<typeof activeExecutionSchema>;
export type ExecutionCompleted = z.infer<typeof executionCompletedSchema>;

export function createEmptyJob(): Job {
  return {
    id: crypto.randomUUID(),
    name: "Novo fluxo",
    sourcePath: "",
    targetPath: "",
    mode: "copy",
    conflict: "skip",
    filters: fileFilterSchema.parse({}),
    options: transferOptionsSchema.parse({}),
    schedule: null,
    scripts: scriptsConfigSchema.parse({}),
    notify: notifyConfigSchema.parse({}),
    enabled: true,
    lastRun: null,
    nextRun: null,
  };
}

export function countActiveFilters(filters: FileFilter): number {
  let n = 0;
  if (filters.includeExtensions.length) n++;
  if (filters.excludePatterns.length || filters.excludePresetIds.length) n++;
  if (filters.nameRegex || filters.pathRegex) n++;
  if (filters.minSizeBytes != null || filters.maxSizeBytes != null) n++;
  if (filters.maxDepth != null) n++;
  if (filters.modifiedAfter || filters.modifiedBefore || filters.createdAfter || filters.createdBefore || filters.olderThanDays) n++;
  if (filters.contentContains) n++;
  if (!filters.recursive) n++;
  if (filters.includeHidden) n++;
  if (filters.symlinkMode !== "follow") n++;
  if (filters.skipEmptyFiles) n++;
  return n;
}

File: crates/autoflow-domain/src/job.rs
use chrono::{DateTime, Utc};

use serde::{Deserialize, Serialize};

use uuid::Uuid;



use crate::{

    DomainError, FileFilter, NotifyConfig, ScheduleConfig, ScriptsConfig, TransferOptions,
    WatchConfig,
};



#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]

#[serde(rename_all = "camelCase")]

pub struct Job {

    pub id: Uuid,

    pub name: String,

    pub source_path: String,

    pub target_path: String,

    pub mode: JobMode,

    pub conflict: ConflictStrategy,

    #[serde(default)]

    pub filters: FileFilter,

    #[serde(default)]

    pub options: TransferOptions,

    #[serde(default)]
    pub schedule: Option<ScheduleConfig>,

    #[serde(default)]
    pub watch: Option<WatchConfig>,

    #[serde(default)]
    pub scripts: ScriptsConfig,

    #[serde(default)]

    pub notify: NotifyConfig,

    pub enabled: bool,

    pub last_run: Option<DateTime<Utc>>,

    pub next_run: Option<DateTime<Utc>>,

    /// Legacy field migrated into `filters` by `normalize()`.

    #[serde(default, skip_serializing)]

    include_extensions: Vec<String>,

    #[serde(default, skip_serializing)]

    recursive: Option<bool>,

}



#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]

#[serde(rename_all = "camelCase")]

pub struct JobSummary {

    pub id: Uuid,

    pub name: String,

    pub source_path: String,

    pub target_path: String,

    pub mode: JobMode,

    pub enabled: bool,

    pub last_run: Option<DateTime<Utc>>,

    pub next_run: Option<DateTime<Utc>>,

    pub schedule_enabled: bool,

    pub watch_enabled: bool,

}



#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]

#[serde(rename_all = "lowercase")]

pub enum JobMode {

    Copy,

    Move,

}



#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]

#[serde(rename_all = "lowercase")]

pub enum ConflictStrategy {

    Skip,

    Overwrite,

    Rename,

}



impl Job {

    pub fn new(name: impl Into<String>, source_path: impl Into<String>, target_path: impl Into<String>) -> Self {

        Self {

            id: Uuid::new_v4(),

            name: name.into(),

            source_path: source_path.into(),

            target_path: target_path.into(),

            mode: JobMode::Copy,

            conflict: ConflictStrategy::Skip,

            filters: FileFilter::default(),

            options: TransferOptions::default(),

            schedule: None,

            watch: None,

            scripts: ScriptsConfig::default(),

            notify: NotifyConfig::default(),

            enabled: true,

            last_run: None,

            next_run: None,

            include_extensions: Vec::new(),

            recursive: None,

        }

    }



    pub fn normalize(&mut self) {

        if self.filters.include_extensions.is_empty() && !self.include_extensions.is_empty() {

            self.filters.include_extensions = std::mem::take(&mut self.include_extensions);

        }

        if let Some(recursive) = self.recursive {

            self.filters.recursive = recursive;

        }

        self.recursive = None;

        self.include_extensions.clear();

        let watch_active = self.watch.as_ref().is_some_and(|w| w.enabled);

        if watch_active {
            if let Some(schedule) = &mut self.schedule {
                schedule.enabled = false;
            }
        } else if self.schedule.as_ref().is_some_and(|s| s.enabled) {
            if let Some(watch) = &mut self.watch {
                watch.enabled = false;
            }
        }

        if watch_active {
            self.next_run = None;
        } else if let Some(schedule) = &self.schedule {

            if schedule.enabled {

                self.next_run = schedule.compute_next_run(Utc::now());

            } else {

                self.next_run = None;

            }

        } else {

            self.next_run = None;

        }

    }



    pub fn validate(&self) -> Result<(), DomainError> {

        let name = self.name.trim();

        if name.is_empty() || name.len() > 120 {

            return Err(DomainError::Validation(

                "name must be between 1 and 120 characters".into(),

            ));

        }

        if self.source_path.trim().is_empty() {

            return Err(DomainError::Validation("source_path is required".into()));

        }

        if self.target_path.trim().is_empty() {

            return Err(DomainError::Validation("target_path is required".into()));

        }

        if let Some(schedule) = &self.schedule {

            schedule.validate()?;

        }

        if let Some(watch) = &self.watch {

            watch.validate()?;

        }

        self.scripts.validate()?;

        Ok(())

    }



    pub fn summary(&self) -> JobSummary {

        JobSummary {

            id: self.id,

            name: self.name.clone(),

            source_path: self.source_path.clone(),

            target_path: self.target_path.clone(),

            mode: self.mode,

            enabled: self.enabled,

            last_run: self.last_run,

            next_run: self.next_run,

            schedule_enabled: self.schedule.as_ref().is_some_and(|s| s.enabled),

            watch_enabled: self.watch.as_ref().is_some_and(|w| w.enabled),

        }

    }

}



impl Default for Job {

    fn default() -> Self {

        Self::new("Novo fluxo", "", "")

    }

}


